harness/determined/pytorch/_reducer.py

```diff
@@ -41,3 +42,146 @@ def _reduce_metrics(
         return np.min(metrics)
     else:
         raise NotImplementedError
+
+
+class MetricReducer(metaclass=abc.ABCMeta):
+    """
+    Efficiently aggregating validation metrics across a multi-slot distributed evaluation is done
+    in three steps:
+
+    1. Gather all the values to be reduced during the reduction window (either a training or a
+       validation workload).  In a multi-slot trial, this is done on each slot independently.
+
+    2. Calculate the initial per-slot reduction.  This step is optional, but can be important to
+       reducing network communication costs, as the output of this step will be passed over the
+       network to other slots before Step 3.
+
+    3. Reduce the per-slot reduction values from Step 2 into a final metric.
+
+    The MetricReducer API makes it possible for users to define a maximally efficient custom metric
+    by exposing these steps to users:
+
+       -  Step 1 is fully defined by the user; it is not part of the interface.  This flexibility
```

Comment:
"executed" and "defined" are not interchangable, and this is saying that the user is responsible for actually defining the behavior.

This documentation is actually about the interface, and `MetricReducer.update()` is definitely not part of the interface, so it would be wrong to say the user is responsible for calling `.update()`.

Reply:
"defined" sounds like implementing a method interface to me. So I wanted to use a different word but `defined` isn't very bad since we have a code example.